client/src/translations/index.tsx

```diff
@@ -0,0 +1,790 @@
+    if (data) {
+      setLastData(data);
+    }
+  }, [data]);
+
+  if (locale.toLowerCase() === "en-us") {
+    return (
+      <Container>
+        {!dataLocales && !errorLocales && <Loading estimate={2000} />}
+        {errorLocales && (
+          <div className="error-message">
+            <h3>Server error</h3>
+            <pre>{errorLocales.toString()}</pre>
+          </div>
+        )}
+        {dataLocales && <ShowLocales locales={dataLocales.locales} />}
+      </Container>
+    );
+  }
+
+  const lastStorageData = getStorage(locale);
+
+  return (
+    <Container>
+      {lastData && !error && isValidating && (
+        <p style={{ float: "right" }}>Reloading...</p>
+      )}
+      {!data && !error && !lastData && (
+        <Loading
+          estimate={
+            (lastStorageData && lastStorageData.lastLoadTime) || 30 * 1000
+          }
+        />
+      )}
+      {error && <ShowSearchError error={error} />}
+      {lastData && (
+        <div className="filter-documents">
+          <FilterControls />
+          <DocumentsTable
+            counts={lastData.counts}
+            documents={lastData.documents}
+          />
+        </div>
+      )}
+      {data && <BuildTimes times={data.times} />}
+    </Container>
+  );
+}
+
+function Container({ children }: { children: React.ReactNode }) {
+  return (
+    <div className="all-translations">
+      <PageContentContainer>{children}</PageContentContainer>
+    </div>
+  );
+}
+
+function ShowLocales({ locales }: { locales: Locale[] }) {
+  return (
+    <div>
+      <h2>Select a locale</h2>
+      <ul>
+        {locales.map((locale) => {
+          return (
+            <li key={locale.locale}>
+              <Link to={`/${locale.locale}/_translations`}>
+                {locale.language.English} ({locale.locale})
+              </Link>{" "}
+              {!locale.isActive && <small>not active</small>}
+            </li>
+          );
+        })}
+      </ul>
+    </div>
+  );
+}
+
+function ShowSearchError({ error }) {
+  return (
+    <div className="error-message search-error">
+      <h3>Search error</h3>
+      <pre>{error.toString()}</pre>
+    </div>
+  );
+}
+
+function BuildTimes({ times }: { times: Times }) {
+  function format(ms: number) {
+    if (ms > 1000) {
+      const s = ms / 1000;
+      return `${s.toFixed(1)} seconds`;
+    } else {
+      return `${Math.trunc(ms)} milliseconds`;
+    }
+  }
+  return (
+    <div className="search-times">
+      <p>Time to find all documents {format(times.took)}</p>
+    </div>
+  );
+}
+
+function Loading({ estimate }: { estimate: number }) {
+  const startLoadingTime = new Date();
+  const [estimateEndTime] = React.useState(
+    new Date(startLoadingTime.getTime() + estimate)
+  );
+
+  const INTERVAL_INCREMENT = 700;
+  const [elapsed, setElapsed] = React.useState(0);
+  React.useEffect(() => {
+    const interval = setInterval(() => {
+      setElapsed((state) => state + INTERVAL_INCREMENT);
```

Comment:
`setInterval` is not guaranteed to respect the given interval, I think most browsers run it at most once-per-second when the tab is inactive and their might also be other strategies in place on mobile or in powersave mode. Hence I'd recommend doing `performance.now()` based elapsed time calculation (though I guess exact precision is not super important here)

Reply:
`performance.now` or not, I think I know what you're saying. This current implementation relies on `setInterval` being "accurate". It would be better to just trigger a hook change, which then triggers a re-render and the re-render looks at `new Date()` vs. `startTime`.

Reply:
I rewrote it so that the `setInterval` always clicks after 1,000ms. But the "elapsed" time isn't based on that, it's based on comparing `new Date()` with the initial `new Date()` when the component was created. 
Now, `setInterval` just increments an integer which triggers a re-render, roughly, every 1s.